lua/tests/infoview_lsp_spec.lua

```diff
@@ -0,0 +1,62 @@
+local infoview = require('lean.infoview')
+local helpers = require('tests.helpers')
+
+local function get_info_lines(infoview_info)
+  return table.concat(vim.api.nvim_buf_get_lines(infoview_info.bufnr, 0, -1, true), "\n")
+end
+
+local function infoview_lsp_test(pos, expected)
+    local infoview_info = infoview.open()
+    vim.api.nvim_win_set_cursor(0, pos)
+    infoview.update(infoview_info.bufnr)
+    local result, _ = vim.wait(5000, function()
+      for _, string in pairs(expected) do
+        local curr = get_info_lines(infoview_info)
+        if not curr:find(string) then return false end
+      end
+      return true
+    end)
+    assert.message( "expected: " .. vim.inspect(expected) ..  ", actual: " .. get_info_lines(infoview_info)
+    ).is_true(result)
+end
+
+describe('infoview', function()
+  helpers.setup {
+    infoview = { enable = true },
+    lsp = { enable = true },
+    lsp3 = { enable = true },
+  }
+
+  it('lean 3', function()
+    vim.api.nvim_command("edit lua/tests/fixtures/example-lean3-project/test.lean")
+    helpers.lsp_wait()
+    vim.wait(5000)
```

Comment:
Seems like a long time to wait unconditionally here, what are we waiting for, the message to update?

Can we just do the assertion in the wait so it passes as soon as we find the match?

Reply:
Not entirely sure what we're waiting for here, but if we request the `hover`/`plainGoal` data immediately without waiting it returns empty. I suspect we're waiting for the language server to finish a pass over the file, @gebner does that sound right?

You can try this out manually as well - when you first start things up, you'll notice that the infoview doesn't work for the first few seconds, regardless of if you're moving the cursor around (so it's not #60).

In Lean 4 there's the `processing...` diagnostic that is sent down the file to indicate progress - but rather than immediately implementing a check for that in these tests, I think we should wait until we've implemented that in the actual code (for a message in the infoview that indicates whether or not the server has processed up to the current line, similar to what is done in VSCode), which we can then use here.

Reply:
So can we just condition on the infoview being nonempty here?

Reply:
Think I fixed it as you've suggested - let me know what you think.

Reply:
> Not entirely sure what we're waiting for here, but if we request the `hover`/`plainGoal` data immediately without waiting it returns empty. I suspect we're waiting for the language server to finish a pass over the file, @gebner does that sound right?

Yes, the requests always return the "current state".  If Lean hasn't processed that part of the file yet, you get an empty response.

You could try the (internal and completely standards-violating) `textDocument/waitForDiagnostics` request, which is supposed to block until Lean is finished.